services/circleci/circleci.service.js

```diff
@@ -0,0 +1,106 @@
+'use strict'
+
+const Joi = require('joi')
+const BaseJsonService = require('../base-json')
+
+const circleSchema = Joi.array()
+  .items(Joi.object({ status: Joi.string() }))
+  .min(1)
+  .required()
+
+module.exports = class CircleCi extends BaseJsonService {
+  async fetch({ token, type, userRepo, branch }) {
+    let url = `https://circleci.com/api/v1.1/project/${type}/${userRepo}`
+    if (branch != null) {
+      url += `/tree/${branch}`
+    }
+    const query = { filter: 'completed', limit: 1 }
+    if (token) {
+      query['circle-token'] = token
+    }
+    return this._requestJson({
+      url,
+      schema: circleSchema,
+      options: { qs: query },
+      errorMessages: { 404: 'project not found' },
+    })
+  }
+
+  static render({ status, color }) {
+    return { message: status, color: color }
+  }
+
+  transform(data) {
+    let passCount = 0
+    let circleStatus, shieldsStatus, color
+
+    for (let i = 0; i < data.length; i++) {
+      circleStatus = data[i].status
+      if (['success', 'fixed'].includes(circleStatus)) {
+        passCount++
+      } else if (circleStatus === 'failed') {
+        shieldsStatus = 'failed'
+        color = 'red'
+        return { status: shieldsStatus, color: color }
+      } else if (['no_tests', 'scheduled', 'not_run'].includes(circleStatus)) {
+        color = 'yellow'
+        shieldsStatus = circleStatus.replace('_', ' ')
+        return { status: shieldsStatus, color: color }
+      } else {
+        color = 'lightgrey'
+        shieldsStatus = circleStatus.replace('_', ' ')
+        return { status: shieldsStatus, color: color }
```

Comment:
So it seems the purpose of this function is to summarize the multiple statuses down to a single status. I'm fine with that, though it does seem possible that it's better thought of as a rendering concern. For example, we could have a badge option that showed these like **2 passed, 1 failed**.

Also, this looks buggy. What if the first test is scheduled and the second is failed? That should be reported as failed, but I think it will show as scheduled instead. Probably it would be better to check if any of the statuses are `'failed'`, then move down the list.

Would it be cleaner (and less buggy) to return an array of statuses, or extract an array of statuses in `handle()`, and delegate the summarizing and coloring to `render()`?

Reply:
This is a very good point which I did not consider when I was looking at #1815 but now I think about this, there are many worms in this can..

It is clear that if *all* tests are passing, that should summarise to 'pass', and if *any* of the tests are failing, that should summarise to 'fail', but if we get back something like:

```
[
  { 'status': 'running' },
  { 'status': 'infrastructure_fail' },
  { 'status': 'scheduled' },
]
```

or

```
[
  { 'status': 'success' },
  { 'status': 'success' },
  { 'status': 'queued' },
  { 'status': 'no_tests' },
]
```

what is an appropriate summary of that?

Looking at [the docs](https://circleci.com/docs/api/v1-reference/), it seems that we could get any arbitrary combination of:

* retried
* canceled
* infrastructure_fail
* timedout
* not_run
* running
* failed
* queued
* scheduled
* not_running
* no_tests
* fixed
* success

Reply:
```
[
  { 'status': 'running' },
  { 'status': 'infrastructure_fail' },
  { 'status': 'scheduled' },
]
```

-> **infrastructure_fail**

```
[
  { 'status': 'success' },
  { 'status': 'success' },
  { 'status': 'queued' },
  { 'status': 'no_tests' },
]
```

Huh, is that actually possible? **no tests** seems to mean there would be no other results. If that ever happened I'd be okay displaying **queued** or **indeterminate**.

In the case of

```
[
  { 'status': 'success' },
  { 'status': 'success' },
  { 'status': 'no_tests' },
]
```

I think **indeterminate** is appropriate.

Maybe we could prioritize the results by severity, and display the highest severity result that appears? We'll have to make some judgement calls when we do that, though I think that's okay.

We could special-case **no tests** which really should only exist on its own.

Reply:
I think you might have taken those examples too literally - I just made up some arbitrary combinations :)

However, its very useful that this has prompted me to think about this more clearly and properly read the documentation because it turns out:

a) I've been trying to solve completely the wrong problem
b) It is only through luck that #1815 has not actually broken anything for users

What I _thought_ I've been doing here is combining the statuses of a list of jobs/workflows to make a build status, but what I've _actually_ been trying to do is combine >1 build statuses (which already summarise multipe jobs/workflows) to make.. a thing that makes no sense. By sheer good luck, we were passing `&limit=1` the whole time so in the real world it all made no difference anyway because we've only ever really been operating on an array of length 1 (despite having written some tests which assume an array of length >1).

So.. the upshot of all that is:
* #1815 was completely pointless and fixed nothing - I've put the code back how it was before I started in 90160bc
* I've also added validation and a test to clarify that we are expecting an array of length 1 and removed the pointless tests in 90160bc
* Whatever the issue was in #1792 is not fixed and we should re-open that issue. I was completely wrong about the problem.

:man_facepalming:

Reply:
I agree with much of that, though I was just looking at https://circleci.com/docs/api/v1-reference/ and it seems to me that a build does _not_ combine multiple jobs. A build seems to refer to e.g. a single tile on https://circleci.com/gh/badges/shields = an individual job.

https://circleci.com/api/v1.1/project/github/badges/shields

Unless `outcome` and `status` are conveying different things, or there's something else I'm not seeing…

Reply:
I feel like `outcome` may actually be the one we are looking for:
```javascript
  "outcome" : "failed", // :canceled, :infrastructure_fail, :timedout, :failed, :no_tests or :success
  "status" : "failed", // :retried, :canceled, :infrastructure_fail, :timedout, :not_run, :running, :failed, :queued, :scheduled, :not_running, :no_tests, :fixed, :success
```

Reply:
Right, yep so we are only looking at one job when there may be N jobs in the build (although I think we are using using the wrong terminology there as the docs refer to one record in this array as a 'build'). I'm going to refer to 'build' and 'group' moving forwards.

`status` and `outcome` seem to both relate to a build. For example, if we look at the branch [dependabot/npm_and_yarn/simple-icons-1.8.4](https://circleci.com/api/v1.1/project/github/badges/shields/tree/dependabot/npm_and_yarn/simple-icons-1.8.4) `build_num`s 11388, 11387, 11386, 11385 and 11384 all group together, but if we look at the status and outcome values:

```
build_num    status    outcome
11388        fixed     success
11387        failed    failed
11386        success   success
11385        failed    failed
11384        failed    failed
```

Although 'outcome' is probably useful to reduce the number of possible values we have to deal with, if it was summarising the group, I'd expect them all to be 'failed'.

So we do need to aggregate multiple status (or outcome values). What I'm slightly struggling with now is how to connect builds in an array like [this](https://circleci.com/api/v1.1/project/github/badges/shields/tree/master) (which contains builds relating to multiple groups) as being part of the same group (which is what we're trying to summarise).

One thing I have thought of is we have the `vcs_revision` property, so we can identify builds in the array which were run against the same commit (on the same branch). Is that a solid way of doing this, or can we run the same build more than once on the same commit? I _think_ if we restart a build it re-uses the same number.

Reply:
Looking at an example, picking all of the objects with the same `vcs_revision` is going to double-count retried builds. For example:
https://circleci.com/gh/badges/shields/11764 and
https://circleci.com/gh/badges/shields/11769
are both `services-pr` on the same commit hash on the same branch.

Reply:
Can we use `workflows.workflow_id`?

```js
workflows: {
  job_name: "frontend",
  job_id: "ef27df67-7aad-4022-a159-2c55718da667",
  workflow_id: "5b94f353-2060-4c80-8b35-941174632ce3",
  workspace_id: "5b94f353-2060-4c80-8b35-941174632ce3",
  upstream_job_ids: [
    "4fac3028-db01-4497-a3c5-97c75194eb0d"
  ],
  upstream_concurrency_map: { },
  workflow_name: "on-commit"
}
```

Reply:
Yeah, i think that is going to be the only way, although we are also going to need to increase the `?limit=`